Src/DependencyCollector/Shared/Implementation/TelemetryDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,293 @@
+                }
+                this.individualSubscriptions.Add(subscription);
+            }
+        }
+
+        /// <summary>
+        /// Notifies the observer that the provider has finished sending push-based notifications.
+        /// <seealso cref="IObserver{T}.OnCompleted()"/>
+        /// </summary>
+        public void OnCompleted()
+        {
+        }
+
+        /// <summary>
+        /// Notifies the observer that the provider has experienced an error condition.
+        /// <seealso cref="IObserver{T}.OnError(Exception)"/>
+        /// </summary>
+        /// <param name="error">An object that provides additional information about the error.</param>
+        public void OnError(Exception error)
+        {
+        }
+
+        public void Dispose()
+        {
+            if (this.individualSubscriptions != null)
+            {
+                foreach (var individualSubscription in this.individualSubscriptions)
+                {
+                    individualSubscription.Dispose();
+                }
+            }
+
+            if (this.listenerSubscription != null)
+            {
+                this.listenerSubscription.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Handler for Activity stop event (response is received for the outgoing request).
+        /// </summary>
+        /// <param name="diagnosticListener">The diagnostic source of the activity.</param>
+        internal void OnActivityStop(DiagnosticListener diagnosticListener)
+        {
+            Activity currentActivity = Activity.Current;
+            if (currentActivity == null)
+            {
+                DependencyCollectorEventSource.Log.CurrentActivityIsNull();
+                return;
+            }
+
+            DependencyCollectorEventSource.Log.TelemetryDiagnosticSourceListenerActivityStopped(currentActivity.Id, currentActivity.OperationName);
+
+            // extensibility point - can chain more telemetry extraction methods here
+            ITelemetry telemetry = this.ExtractDependencyTelemetry(currentActivity);
+            if (telemetry == null)
+            {
+                return;
+            }
+
+            // properly fill dependency telemetry operation context
+            telemetry.Context.Operation.Id = currentActivity.RootId;
+            telemetry.Context.Operation.ParentId = currentActivity.ParentId;
+            telemetry.Timestamp = currentActivity.StartTimeUtc;
+
+            telemetry.Context.Properties["DiagnosticSource"] = diagnosticListener.Name;
+            telemetry.Context.Properties["Activity"] = currentActivity.OperationName.Substring(0, currentActivity.OperationName.Length - StopSuffix.Length);
+
+            this.client.Initialize(telemetry);
+
+            this.client.Track(telemetry);
+        }
+
+        internal DependencyTelemetry ExtractDependencyTelemetry(Activity currentActivity)
+        {
+            DependencyTelemetry telemetry = new DependencyTelemetry();
```

Comment:
based on the span.kind we can decide whether it's request or dependency

Reply:
do you mean "span.kind"? ok, I can inspect it